clients/src/main/java/org/apache/kafka/common/protocol/Errors.java

```diff
@@ -495,8 +497,22 @@ public ApiException build(String message) {
             public ApiException build(String message) {
                 return new OperationNotAttemptedException(message);
             }
-        });
-
+    }),
+    KAFKA_STORAGE_ERROR(56, "Disk error when trying to access log file on the disk.",
+        new ApiExceptionBuilder() {
+            @Override
+            public ApiException build(String message) {
+                return new KafkaStorageException(message);
+            }
+    }),
+    UNKNOWN_ERROR_CODE(57, "The client received an unknown error code when processing the response. This may happen if " +
+            "the client library version is lower than the server library version",
+        new ApiExceptionBuilder() {
```

Comment:
Another way I look at this is that, regardless of what the underlying code does, the outside code, i.e. those methods that determines the response, should catch all exception and determine the error in the response properly. This is similar to why we currently catch `Throwable`. It is just that, instead of catching IOException (as part of the throwable) and return UnknownServerException, it seems more reasonable to call `maybeAddLogDirFailure()` and return KafkaStorageException on IOException.